aws_tools/s3_handler.py

```diff
@@ -30,14 +30,15 @@ def download_file(self, key, local_file):
     # Downloads all the files in S3 that have a prefix of `key_prefix` from `bucket` to the `local` directory
     def download_dir(self, key_prefix, local):
         paginator = self.client.get_paginator('list_objects')
-        result = paginator.paginate(Bucket=self.bucket_name, Delimiter='/', Prefix=key_prefix)
-        for subdir in result['CommonPrefixes']:
-            self.download_dir(subdir['Prefix'], local)
-        for f in result['Contents']:
-            dir = local + os.sep + f['Key']
-            if not os.path.exists(os.path.dirname(dir)):
-                os.makedirs(os.path.dirname(dir))
-            self.download_file(f['Key'], dir)
+        for result in paginator.paginate(Bucket=self.bucket_name, Delimiter='/', Prefix=key_prefix):
+            if result.get('CommonPrefixes') is not None:
+                for subdir in result.get('CommonPrefixes'):
+                    self.download_dir(subdir.get('Prefix'), local)
+            if result.get('Contents') is not None:
```

Comment:
Should be as it is, as it's dealing with directories in the first if (actually shouldn't even need the if, but just to make sure that the results has the CommonPrefixes), and the second deals with files. If neither, I'm not sure what it's trying to download! Directories and files are all its getting, and for each it either doing a download_dir for a dir, or download_file the file.